### PR TITLE
[BugFix][SpecDecode] Fix eagle-family aivec crash: blocking H2D for backup_next_token_ids

### PR DESCRIPTION
**Pull request by DonJose2002**
### What this PR does / why we need it?
Eagle-family speculative decoding on Ascend dies with an engine-fatal hardware exception at longer prompts:

```
EE9999: vector core exception (aivec), MTE accesses an invalid GM address
fault kernel (plog): GatherV2_..._high_precision   # a vocab-sized embedding gather
args: vocab size 151936; index bytes carry 0xa5a5a5a5  # CANN's uninit-memory poison marker
```

100% reproducible at 16K prompts (Qwen3-8B + EAGLE3 head, K=5), intermittent at 4K, across multiple physical cards; all in-flight requests are truncated when the engine dies.

Root cause (bisected on 910B3 with timing-preserving probes): `prepare_next_token_ids_padded` ships the fallback `backup_next_token_ids` via `CpuGpuBuffer.copy_to_gpu` → `gpu.copy_(cpu, non_blocking=True)`. On torch_npu, pinned H2D copies ride the SDMA engine, whose ordering w.r.t. subsequent compute kernels on the launch stream is **not guaranteed**. When the race is lost, the `torch.where` consumer reads uninitialized memory and the garbage ids scatter into the drafter's `input_ids`; the vocab-sized embedding gather then indexes out of bounds and faults the vector core.

Why upstream GPU is safe: `cudaMemcpyAsync` enqueued on a stream is strictly ordered w.r.t. later kernels on that stream — the same `non_blocking=True` copy is safe on CUDA. The guarantee does not carry over to torch_npu's SDMA path; NPU ports cannot inherit CUDA stream-ordering intuitions.

Why eagle-family only: ngram/suffix drafters run after bookkeeping on CPU lists and never call this path (structurally immune). All other padded methods — dflash/dspark (via `use_eagle()`) and draft_model — DO share this racing copy and are narrower-window instances of the same bug, not immune; this fix covers them too. The frequency difference is trial counts, not immunity: we later measured (device-side counters, zero host sync) that the `torch.where` fallback-selection gate opens only ~7 rows per ~845 steps (request-boundary steps where the backup carries a real token), and the racing copy loses exactly once per run there — that single stale value is the engine-killing event; eagle's µs-dense launch cadence simply widens the launch-ahead window far more than the heavier draft_model/dflash steppers.

Fix: blocking copy for this <1KB payload (`num_reqs` int32). The microsecond cost buys a deterministic ordering. `CpuGpuBuffer.copy_to_gpu` is deliberately left untouched — larger buffers legitimately rely on non_blocking performance; this is a targeted fix at the racing consumer.
### Does this PR introduce _any_ user-facing change?
No config/API changes. Previously-crashing workloads (eagle/eagle3/mtp + longer prompts) now complete; numerics untouched (the racing copy only feeds fallback ids for requests whose sampled token was discarded).
### How was this patch tested?
Race conditions cannot be regression-tested on CPU; the evidence chain is the test:
- Repro: 16K single request, eagle3 + Qwen3-8B, K=5 — unpatched fails 4/4 configurations (clean empty streams / HTTP 500 → engine death); patched passes 4/4.
- Attribution probes (env-gated, research branch only): `ASCEND_LAUNCH_BLOCKING=1` rescues the crash (race confirmation); per-tensor clamp probes that preserve timing bisect the carrier — clamp only `next_token_ids` passes 16K, clamp only `target_token_ids` does not.
- Value-path verdict (on-device counters, zero host sync, read out via SIGUSR1 at end of a rescued run): across 843 steps the `torch.where` fallback gate opened 7 times (request-boundary rows, where the backup legitimately carries a real token via `get_token_id`), and exactly ONE out-of-vocab id escaped the `where` (c3=1). A value-capture probe pinned the escaped value at exactly **-1** (esc=[-1,-1]): the backup buffer's stale sentinel — upstream `get_token_id` returns -1 for not-yet-committed positions, so every normal step leaves -1 in the buffer, and the racing `where` on a gate-open step can observe the pre-landing (sentinel) state instead of the freshly written real token. A parallel counter excluding the sentinel read zero (c1x=0): no other garbage, no CANN poison marker ever entered the value path. That one -1 escape is precisely the unpatched engine's fatal event: the id scatters into the drafter's `input_ids` and the vocab-sized embedding gather indexes GM out of bounds. The run survived only because the clamp probe sanitized it after the counters. With the fix active the same 847-step run shows c3=0 (copy always lands before the `where`). All links of the value chain measured.
- Full matrix post-fix: 9/9 cells (4K/16K/32K × concurrency 1/4/16) all green, accept length 2.34–2.66 (counters and burst-gap estimates agree), where the unpatched engine never survived the 16K row.
- Hardware: Ascend 910B3, v0.23.0 docker, reproduced across 4 different physical cards.

> Note: developed with AI assistance (GLM-5.3); human-reviewed, human-validated on hardware.
- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 1]
```diff
@@ -1780,7 +1780,17 @@ def prepare_next_token_ids_padded(
...
+        # pinned H2D rides the SDMA engine, whose ordering w.r.t. subsequent
+        # compute kernels on the launch stream is not guaranteed. When the
+        # race is lost, the torch.where below consumes uninitialized memory
+        # (0xa5a5a5a5 marker) and the garbage ids reach the drafter's
+        # vocab-sized embedding gather, faulting the vector core (aivec MTE
+        # invalid GM address, EE9999, engine death). Payload is num_reqs
+        # int32 (<1KB) - a blocking copy is effectively free.
+        self.backup_next_token_ids.gpu[:num_reqs].copy_(
+            self.backup_next_token_ids.cpu[:num_reqs], non_blocking=False
+        )
```
[the remark this answers is not in the document]
Comment (DonJose2002) [reply vllm_ascend/spec_decode/llm_base_proposer.py:1793]: Thanks for the review! The concern is reasonable on its face, but the premise doesn't hold for this function, and the proposed change would guard a path that is architecturally dead. Three legs:

**1. `batch_size == num_reqs` here — the "padded" in the name is spec-token padding, not request padding.**

The caller contract (see the comment at the call site in `model_runner_v1.py`, "When using padded-batch, the sampled_token_ids should be the gpu tensor of sampled tokens for each request, of shape `(num_reqs, num_spec_tokens + 1)` with rejected tokens having value -1") is that `sampled_token_ids` is `(num_reqs, K+1)`: each *request's row* is padded to K+1 spec-token slots (rejected positions = -1), while the request dimension itself is not padded — it is exactly `gpu_input_batch.num_reqs`. So `batch_size = sampled_token_ids.shape[0] == num_reqs`, and the `self.backup_next_token_ids.gpu[:batch_size]` slice read by the `torch.where` is precisely the slice the blocking copy writes. There is no `[num_reqs:batch_size]` gap in this function.

(Request-count padding *does* exist elsewhere — e.g. FULL-graph drafter batches pad token counts up to capture buckets — but that happens downstream at the drafter-input level, after this function returns.)

**2. Even where the drafter runs a padded batch, phantom rows never source their ids from `next_token_ids`.**

- Eagle branch: `self.input_ids[token_indices_to_sample] = next_token_ids` scatters exactly `num_reqs` entries — `token_indices_to_sample` is derived from the (unpadded) `query_start_loc`, one index per real request.
- CopyAndExpand branch (draft_model / dflash / dspark): `npu_copy_and_expand_eagle_inputs` consumes `next_token_ids` under real `query_start_loc`/`query_end_loc` boundaries and writes phantom/unused slots with its explicit `padding_token_id=0` argument.

So the drafter's embedding gather never indexes with a value that originated in the tail of `backup_next_token_ids` — on any path.

**3. Empirically, stale tails existed before this PR and never crashed anything.**

The pre-fix code shipped the **full** pinned buffer every step (`copy_to_gpu()` with no size argument), so the GPU slots beyond `num_reqs` have always held stale values from earlier, larger batches — including `-1` sentinels. If those slots were consumed by the gather, the unpatched engine would have crashed on them constantly; it did not. The only escape we ever measured (device-side counters, `c3=1`, value-capture `esc=[-1,-1]`) was on a **real request-boundary row** via the copy-vs-`where` race — which is exactly what the blocking copy eliminates (fix-active counters: `c3=0`; 9/9 matrix incl. concurrency-16 batch-shrink scenarios green).

Given the above, copying `[:batch_size]` with a zero-padded CPU tail would be harmless but strictly additive: extra host writes and a larger blocking copy to refresh slots whose values are dead by construction. We'd rather keep the fix minimal. Happy to revisit if a future consumer of the tail emerges — at that point the right shape would probably be an explicit invariant on the whole buffer, not a per-step full copy.

**Comment by DonJose2002** [comment]
**Title direction: Correction to our mechanism analysis: repro-level data was an artifact of our own script; the fix is unaffected and more robust than claimed**

Hi maintainers — we need to correct the mechanism analysis in our earlier reply
(and parts of the PR description). Short version: the minimal-reproducer data we
cited turned out to be a **measurement artifact of the reproducer itself**, not
evidence of a platform defect. **The fix in this PR is unaffected, and is more
robust than we originally argued.**

What we found:

1. Our repro used two alternating pinned host buffers (rewritten every 2 steps)
   and its miss counter did not distinguish stale reads from FUTURE reads.
2. A follow-up pure-CANN probe measured that the submission pipeline lets the
   host run ~2048 tasks ahead of the device (dominant lag matched
   2048/tasks-per-step to +-1 across five configurations; overall FIFO order
   holds, with backpressure pacing the host afterwards).
3. With the host far ahead, a late-executing H2D copy reads an
   already-overwritten host buffer and delivers a FUTURE step's value; the
   direction-blind miss counter counted all of it — the entire 0.7%-99.94%
   miss "spectrum" came from this.
4. Single-variable closure (same binary, paired runs): unique per-step host
   slots -> miss=0 in EVERY configuration (incl. 64MiB/step full pressure,
   TQ=0/1, ctypes direct submission); restoring the old alternating design ->
   reproduces the old 99.85%/99.94% exactly, all future-direction.
5. A pure-CANN probe (libascendcl + two-stage libopapi op, same-stream
   copy->AI-core kernel, same pressure, single- and cross-thread) is also
   clean: at the pure-CANN layer, same-stream copy->kernel ordering and data
   visibility are correct.
6. Fence re-check: with unique slots, the event-mode run (side-stream copy +
   stream synchronize) is miss=0/2000 — the sync covers the copy properly;
   the old event-mode miss reproduces only under the old alternating slots
   (0.4%, all future-direction, a startup-transient host run-ahead) and is
   unrelated to fence behavior.

We therefore retract: the "copy is off-stream / fences cannot cover it"
narrative (including the earlier wall-time-based inference), the "CANN
runtime visibility gap" attribution, and the six-cell matrix numbers as
defect evidence.

What still stands: the original bug (eagle-style speculative decoding + 16K
sequences, deterministic crash without the fix, reproduced many times); the
unique escape value -1 (device-side counters, matching the get_token_id
uncommitted-position sentinel feeding gather(-1)); fix efficacy (9/9 config
matrix + all later runs; also insensitive to host-slot design, re-verified
miss=0); and the fix is mechanism-robust — a blocking copy captures the
buffer content synchronously, so later host rewrites can no longer affect
the transfer, which is the right medicine even if the root cause is on the
host-rewrite side.

Root cause now CLOSED (same-day update). A discriminating experiment
(env-gated, no default behavior change): with the race revived, the ONLY
change was the copy's source - a per-step PRIVATE snapshot page instead of
the single pinned page that gets rewritten every step (host-side <1KB
memcpy; the copy itself stays on the original non_blocking path):

- single page (positive control): 0/8, crash - reproduction held;
- 8 private pages: 8/8 green; counters: gate-open count unchanged at 7,
  escapes 1 -> 0 (single-page era: c3=1, esc=[-1,-1]; staged: c3=0,
  esc=none).

Conclusion: the -1 entered through a late-executing async copy reading the
host-rewritten single source page - a use-after-rewrite, NOT an ordering or
visibility problem (if copy->consumer ordering were broken, private pages
could not help and c3 would not be zero). The deep async submission queue
(measured ~2048 tasks) lets the host rewrite an in-flight source page;
eagle's microsecond step loop makes it happen, while on CUDA the host is
pinned by per-step sampling dependencies and the hazard never lands.

The blocking copy in this PR snapshots the page content synchronously -
exactly the right medicine, now with the mechanism proven. A paged
(double-buffered) scheme works equally well and keeps the copy async:
same-day back-to-back runs (16K, concurrency 1) measured out/s
66.1 -> 84.3 (+27.5%), ITL p50 33.7 -> 25.7 ms, acceptance unchanged
(2.48; same-card re-test ruled out card-to-card variance). A possible
follow-up optimization.

Apologies again for the earlier overconfident mechanism narrative — and it
is exactly why this PR keeps the minimal one-line blocking fix, which never
depended on any of those mechanism assumptions and now turns out to be
precisely on target.
